lxd/main_forkusernsexec.go

```diff
@@ -0,0 +1,217 @@
+{
+	int ret;
+	struct statfs sb;
+
+	ret = fstatfs(fd, &sb);
+	if (ret < 0)
+		return false;
+
+	return is_fs_type(&sb, magic_val);
+}
+
+static inline bool drop_groups(void)
+{
+	return setgroups(0, NULL) == 0;
+}
+
+static inline bool switch_resids(uid_t uid, gid_t gid)
+{
+	if (setresgid(gid, gid, gid))
+		return false;
+
+	if (setresuid(uid, uid, uid))
+		return false;
+
+	if (setfsgid(-1) != gid)
+		return false;
+
+	if (setfsuid(-1) != uid)
+		return false;
+
+	return true;
+}
+
+__attribute__ ((noinline)) int __forkusernsexec(void)
+{
+	__do_free_string_list char **argvp = NULL;
+	pid_t pid;
+	ssize_t ret;
+	int fd_socket[2];
+	char c;
+	char *cur;
+
+	if (geteuid() != 0)
+		return log_error(EXIT_FAILURE, "Error: forkexec requires root privileges");
+
+	cur = advance_arg(false);
+	if (!cur || (strcmp(cur, "--help") == 0 ||
+	     strcmp(cur, "--version") == 0 || strcmp(cur, "-h") == 0))
+		return EXIT_SUCCESS;
+
+	if (!fhas_fs_type(FD_PIPE_UIDMAP, PIPEFS_MAGIC))
+		return EXIT_FAILURE;
+
+	if (!fhas_fs_type(FD_PIPE_GIDMAP, PIPEFS_MAGIC))
+		return EXIT_FAILURE;
+
+	ret = socketpair(AF_LOCAL, SOCK_STREAM | SOCK_CLOEXEC, 0, fd_socket);
+	if (ret < 0)
+		return EXIT_FAILURE;
+
+	pid = fork();
+	if (pid < 0)
+		return EXIT_FAILURE;
+
+	if (pid == 0) {
+		int fd_uidmap = -EBADF, fd_gidmap = -EBADF;
+
+		close_prot_errno_disarm(fd_socket[0]);
+
+		ret = read_nointr(fd_socket[1], &c, 1);
+		if (ret != 1)
+			_exit(EXIT_FAILURE);
+
+		ret = splice(FD_PIPE_UIDMAP, 0, fd_uidmap, 0, MAP_SIZE, 0);
+		if (ret < 0 || ret >= MAP_SIZE)
+			_exit(EXIT_FAILURE);
+
+		ret = splice(FD_PIPE_GIDMAP, 0, fd_gidmap, 0, MAP_SIZE, 0);
+		if (ret < 0 || ret >= MAP_SIZE)
+			_exit(EXIT_FAILURE);
+
+		if (write_nointr(fd_socket[1], &c, 1) != 1)
+			_exit(EXIT_FAILURE);
+
+		_exit(EXIT_SUCCESS);
+	}
+
+	close_prot_errno_disarm(fd_socket[1]);
+
+	ret = unshare(CLONE_NEWUSER);
+	if (ret) {
+		kill(pid, SIGKILL);
+		goto out_reap;
+	}
+
+	if (!write_nointr(fd_socket[0], &c, 1)) {
+		kill(pid, SIGKILL);
+		goto out_reap;
+	}
+
+	if (!read_nointr(fd_socket[0], &c, 1)) {
+		kill(pid, SIGKILL);
+		goto out_reap;
+	}
+
+	if (!drop_groups()) {
+		kill(pid, SIGKILL);
+		goto out_reap;
+	}
+
+	if (!switch_resids(0, 0)) {
+		kill(pid, SIGKILL);
+		goto out_reap;
+	}
+
+out_reap:
+	ret = wait_for_pid(pid);
+	if (ret)
+		return EXIT_FAILURE;
+
+	for (char *arg = NULL; (arg = advance_arg(false)); ) {
+		ret = push_vargs(&argvp, arg);
+		if (ret < 0)
+			return log_error(ret, "Failed to add %s to arg array", arg);
+	}
+
+	if (!argvp || !*argvp)
+		return log_error(EXIT_FAILURE, "No command specified");
+
+	execvp(argvp[0], argvp);
+	return EXIT_FAILURE;
+}
+
+void forkusernsexec(void)
+{
+	_exit(__forkusernsexec());
+}
+*/
+import "C"
+
+type cmdForkusernsexec struct {
+	global *cmdGlobal
+}
+
+func (c *cmdForkusernsexec) Command() *cobra.Command {
+	// Main subcommand
+	cmd := &cobra.Command{}
+	cmd.Use = "forkusernsexec <cmd> <args>"
```

Comment:
@brauner @stgraber shouldn't this take 2 FDs for the uid and gid map files?

Reply:
They are at well known locations, it's expecting fd3 and fd4 to be set.

Reply:
Ah OK it would be good to document that in the command help for future users though.

Reply:
This is simply a Go thing not something we do.